### ISSUE TITLE
Port is already occupied when connecting to ElasticSearch

### ISSUE THREAD

**Issue by KrasnovNikita** [jolokia/jolokia#378]
I want to monitor jvm health of Elasticsearch 6x with the help of jolokia java jvm agent 1.5.0. For taking all metrics from Jolokia I use Metricbeat-6.2.3 and then will put all of them to Elasticsearch. Everything is installed on Centos system.
according to this instruction https://jolokia.org/reference/html/agents.html#jvm-agent I do such command:
_su -c "java -jar /tmp/jolokia-jvm-1.5.0-agent.jar --verbose --debug --port=10000 toggle org.elasticsearch.bootstrap.Elasticsearch" -s /bin/bash elasticsearch_
and get such notification
_Couldn't start agent for process matching "org.elasticsearch.bootstrap.Elasticsearch" (PID: 5166)
Possible reason could be that port '10000' is already occupied.
Please check the standard output of the target process for a detailed error message._
And there is no effect when I change a port - still the same problem.

When I try to connect to Tomcat everythings works fine and I get successful response on http://127.0.0.1:8778/jolokia/
What can be the problem of connection to Elastic?

**Comment by rhuss** [comment]
One which UID is your elasticsearch process running ? You need to attach Jolokia with the **very same** UID. root is not good enough.

Actually, it would be much better to configure elasticsearch to start with the jolokia agent from the very beginning, by modifying the startup scripts.

**Comment by marek-sezemsky** [comment]
I've observed same behavior with v2.1.2: `jolokia-*.jar start` report error but subsequent `jolokia-*.jar status` will report success and connectivity to the agent is just fine.

TL;DR - waiting 500ms in 'joloika*.jar start' is probably not enough for some:

- https://github.com/jolokia/jolokia/blob/v1.5.0/agent/jvm/src/main/java/org/jolokia/jvmagent/client/command/StartCommand.java#L52
- https://github.com/jolokia/jolokia/blob/v2.1.2/agent/jvm/src/main/java/org/jolokia/jvmagent/client/command/StartCommand.java#L49

Combination of event timestamps from CLI interaction (akin to `./jolokia.sh start ; while not jolokia-*.jar status $PID: continue`) and JBoss `server.log` events:

	2025-01-08T17:18:07,859992666+00:00 -- CLI 'start' invoked
	2025-01-08T17:18:08,257 DEBUG [com.sun.net.httpserver] (JolokiaStart) HttpServer created http [xxx]
	2025-01-08T17:18:08,261 INFO  [stdout] (JolokiaStart) I> No access restrictor found, access to any MBean is allowed
	2025-01-08T17:18:08,773979090+00:00 -- CLI 'start' reportes error (elapsed ~0:00.89)
	2025-01-08T17:18:08,791321481+00:00 -- CLI 'status 1' invoked
	2025-01-08T17:18:09,284551736+00:00 -- CLI 'status 1' reports error (elapsed ~0:00.47)
	2025-01-08T17:18:09,337309139+00:00 -- CLI 'status 2' invoked
	2025-01-08T17:18:09,735375044+00:00 -- CLI 'status 2' reports error (elapsed ~0:00.37)
	2025-01-08T17:18:09,738 DEBUG [com.sun.net.httpserver] (JolokiaStart) context created: /jolokia/
	2025-01-08T17:18:09,743 INFO  [stdout] (JolokiaStart) Jolokia: Agent started with URL [xxx]
	2025-01-08T17:18:09,750001550+00:00 -- CLI 'status 3' invoked
	2025-01-08T17:18:10,024783728+00:00 -- CLI 'status 3' reports success (elapsed ~0:00.24)

Just from `server.log`, the elapsed time between first and "agent started" messages: elapsed 0:01,486

	2025-01-08T17:18:08,257 DEBUG [com.sun.net.httpserver] (JolokiaStart) HttpServer created http [xxx]
	...
	2025-01-08T17:18:09,743 INFO  [stdout] (JolokiaStart) Jolokia: Agent started with URL [xxx]

On CLI layer timing betwen before-start-invocation and after-status-reported-success: ~0:01,233

	2025-01-08T17:18:08,791321481+00:00 -- CLI 'status 1' invoked
	...
	2025-01-08T17:18:10,024783728+00:00 -- CLI 'status 3' reports success (elapsed ~0:00.24)

What's interesting: when doing cyclic restarts (`while true: status,stop,start,sleep 3s`) this fails occasionally in first few seconds and first few odd starts, however past like 5-7s mark, startups are stable and well-performing at around ~200ms:

	2025-01-08 18:18:15,313 DEBUG [com.sun.net.httpserver] (JolokiaStart) HttpServer created http  [xxx]
	...
	2025-01-08 18:18:15,545 INFO  [stdout] (JolokiaStart) Jolokia: Agent started with URL [xxx]

Recycling EAP server will re-introduce the bug; so this is especially visible in CI test envs with freshly deployed env. Not sure if those delayed startups have deeper meaning or it's just JVM/EAP/GC(?) needing time to "warm up"? For the record HW is well-provisioned with enough CPU, memory and I/O.

In my env this popped up somwhere along the way of bunch of chore updates, including JDK (11 to 17) and jolokia (v1.3.x to v2.1.2): jolokia tests turned from occasional flaky into "start never reliably attaches under 500ms and it's not really about ports".

I tried with -SNAPSHOT where I raised the delay to sth like 1500ms and the 'start' command never ever failed again. So perhaps something like `--startTimeout` might be introduced with resonably increased default (1sec?) and internal polling of agent availability (250ms intervals?) to not wait whole 5s with --startTimeout 5s I'd like to set with it?

I'm open to patching this somehow as I'd love to have this fixed and released soon. :-P

**Comment by grgrzybek** [comment]
Hello and Happy New Year!

I appreciate updating old open issue. Thanks for checking. If increasing the timeout worked for you **and** because hardcoding `500` is never a good idea, then it's obvious we should add this option.
I'm right now working on other Jolokia issue (like IPv6 support/review), so new release is coming soon.

If you have a PR with this command, I'll be happy to review it. Otherwise I'll just add this command before next release somewhere around January 20th.
